aardvark/updater/__init__.py

```diff
@@ -89,75 +79,42 @@ class property ARN filter
-        - Phantomjs exchanges the token for session cookies.
-        - Phantomjs then navigates to the IAM page and executes JavaScript
-        to call GenerateServiceLastAccessedDetails for each ARN.
-        - Every 10 seconds, Phantomjs calls GetServiceLastAccessedDetails
-        - Phantom saves output to a file that is used by `persist()`
-
-        :return: Exit code from phantomjs subprocess32
-        """
-
-        path = os.path.dirname(__file__)
-        console_js = os.path.join(path, 'awsconsole.js')
-
-        with tempfile.NamedTemporaryFile() as f:
-            json.dump(arns, f)
-            f.seek(0)
-            try:
-                p = subprocess32.Popen([
-                    self.current_app.config.get('PHANTOMJS'),
-                    console_js,
-                    token,
-                    f.name,
-                    output_file],
-                    stdout=subprocess32.PIPE, stderr=subprocess32.STDOUT)
-                output, errs = p.communicate(timeout=1200)  # 20 mins
-                self.current_app.logger.debug('Phantom Output: \n{}'.format(output))
-                self.current_app.logger.debug('Phantom Errors: \n{}'.format(errs))
-            except subprocess32.TimeoutExpired:
-                self.current_app.logger.error('PhantomJS timed out')
-                return 1  # return code 1 for timeout
-            except CalledProcessError:
-                self.current_app.logger.error('PhantomJS exited: {}'
-                                              ''.format(p.returncode))
-                return p.returncode
-            else:
-                self.current_app.logger.info('PhantomJS exited: 0')
-                return 0
-
-
-def _get_signin_token(creds):
-    """
-    Exchanges credentials dictionary for a signin token.
-
-    1) Creates URL using credentials dictionary.
-    2) Sends a GET request to that URL and parses the response looking for
-    a signin token.
-
-    :return: Signin Token
-    """
-    url = '{base}?Action=getSigninToken&Session={creds}'
-    url = url.format(base=federation_base_url, creds=creds)
-    return requests.get(url).json()['SigninToken']
+        client = boto3_cached_conn(
+            'iam', account_number=self.account_number, assume_role=self.role_name)
+        return client
+
+    def _call_access_advisor(self, iam, arns):
+        jobs = self._generate_service_last_accessed_details(iam, arns)
+        details = self._get_service_last_accessed_details(iam, jobs)
+        return details
+
+    def _generate_service_last_accessed_details(self, iam, arns):
+        jobs = {}
+        for role_arn in arns:
+            job_id = iam.generate_service_last_accessed_details(Arn=role_arn)['JobId']
+            jobs[role_arn] = job_id
+        return jobs
+
+    def _get_service_last_accessed_details(self, iam, jobs):
+        access_details = {}
+        for role_arn, job_id in jobs.items():
+            while True:
```

Comment:
I'm wondering if there's a better way to do this. Seems like one stuck job here might cause the rest to hang.

Reply:
Sure thing, that should be possible. How would you like to act if something is stuck? A timeout after which some roles yield no result? I didn't review the phantomJS code, did it have some logic for this that we can apply here?